From 6b7e3a7283650cf94892408ca00fb0fd99e880ea Mon Sep 17 00:00:00 2001
From: Matt Riedemann <mriedem@us.ibm.com>
Date: Wed, 18 Mar 2015 12:42:42 -0700
Subject: [PATCH 21/34] Save bdm.connection_info before calling
 volume_api.attach_volume

There is a race in attach/detach of a volume where the volume status
goes to 'in-use' before the bdm.connection_info data is stored in the
database. Since attach is a cast, the caller can see the volume go to
'in-use' and immediately try to detach the volume and blow up in the
compute manager because bdm.connection_info isn't set stored in the
database.

This fixes the issue by saving the connection_info immediately before
calling volume_api.attach_volume (which sets the volume status to
'in-use').

Closes-Bug: #1327218

Change-Id: Ib95c8f7b66aca0c4ac7b92d140cbeb5e85c2717f
(cherry picked from commit 6fb2ef96d6aaf9ca0ad394fd7621ef1e6003f5a1)
---
 nova/tests/compute/test_compute.py   | 2 +-
 nova/tests/virt/test_block_device.py | 2 +-
 nova/virt/block_device.py            | 5 +++++
 3 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/nova/tests/compute/test_compute.py b/nova/tests/compute/test_compute.py
index 9efff99..47e1ac1 100644
--- a/nova/tests/compute/test_compute.py
+++ b/nova/tests/compute/test_compute.py
@@ -573,7 +573,7 @@ class ComputeVolumeTestCase(BaseTestCase):
             })]
             prepped_bdm = self.compute._prep_block_device(
                     self.context, self.instance_object, block_device_mapping)
-            mock_save.assert_called_once_with()
+            self.assertEqual(2, mock_save.call_count)
             volume_driver_bdm = prepped_bdm['block_device_mapping'][0]
             self.assertEqual(volume_driver_bdm['connection_info']['serial'],
                              self.volume_id)
diff --git a/nova/tests/virt/test_block_device.py b/nova/tests/virt/test_block_device.py
index d7fbe73..8b24607 100644
--- a/nova/tests/virt/test_block_device.py
+++ b/nova/tests/virt/test_block_device.py
@@ -378,7 +378,7 @@ class TestDriverBlockDevice(test.NoDBTestCase):
             self.volume_api.attach(elevated_context, fake_volume['id'],
                                    'fake_uuid', bdm_dict['device_name'],
                                    mode=access_mode).AndReturn(None)
-        driver_bdm._bdm_obj.save().AndReturn(None)
+        driver_bdm._bdm_obj.save().MultipleTimes().AndReturn(None)
         return instance, expected_conn_info
 
     def test_volume_attach(self):
diff --git a/nova/virt/block_device.py b/nova/virt/block_device.py
index d2ae79d..773cc50 100644
--- a/nova/virt/block_device.py
+++ b/nova/virt/block_device.py
@@ -261,6 +261,11 @@ class DriverVolumeBlockDevice(DriverBlockDevice):
         if 'data' in connection_info:
             mode = connection_info['data'].get('access_mode', 'rw')
         if volume['attach_status'] == "detached":
+            # NOTE(mriedem): save our current state so connection_info is in
+            # the database before the volume status goes to 'in-use' because
+            # after that we can detach and connection_info is required for
+            # detach.
+            self.save()
             volume_api.attach(context, volume_id, instance.uuid,
                               self['mount_device'], mode=mode)
 
-- 
2.4.3

